Fx DSL improvements + docs

raulraja opened the pull request:
This PR adds more docs for Fx and reduces the number of primtives to:

`fx` A fx monadic continuation block
`effect` lift a suspended side effect into the context of `Fx<F>`
`!effect`apply an effect in a non-blocking fashion binding the result to the left hand side while still suspended in IO.

```kotlin
import arrow.effects.IO
import arrow.effects.extensions.io.fx.fx

suspend fun sayHello(): Unit =
  println("Hello World")
  
suspend fun sayGoodBye(): Unit =
  println("Good bye World!")
  
fun greet(): IO<Unit> =
  fx {
    val pureHello = effect { sayHello() }
    val pureGoodBye = effect { sayHello() }
    !pureHello
    !pureGoodBye
  }
```

pakoito commented:
My Environment/Dispatchers PR is merged too, you may want to use it already.

pakoito commented on modules/core/arrow-typeclasses/src/main/kotlin/arrow/typeclasses/suspended/ApplicativeErrorSyntax.kt:
```diff
@@ -6,27 +6,27 @@ import arrow.core.OptionOf
 
-  suspend fun <A> E.raiseError(): A =
-    run<ApplicativeError<F, E>, Kind<F, A>> { raiseError(this@raiseError) }.bind()
+  suspend fun <A> E.raiseError(): Kind<F, A> =
```
This doesn't need to be suspended anymore, right?

nomisRev commented on modules/effects/arrow-effects-data/src/main/kotlin/arrow/effects/typeclasses/Concurrent.kt:
```diff
@@ -222,15 +222,15 @@ interface Concurrent<F> : Async<F> {
    * @see [arrow.effects.typeclasses.Concurrent.raceN] for a simpler version that cancels losers.
    */
-  fun <A, B, C> raceTriple(ctx: CoroutineContext = dispatchers().default(), fa: Kind<F, A>, fb: Kind<F, B>, fc: Kind<F, C>): Kind<F, RaceTriple<F, A, B, C>>
+  fun <A, B, C> CoroutineContext.raceTriple(fa: Kind<F, A>, fb: Kind<F, B>, fc: Kind<F, C>): Kind<F, RaceTriple<F, A, B, C>>
```
Are we not doing `dispatchers().default().raceTriple` ?

raulraja replied:
We want `NonBlocking.raceTriple` or similar as receiver because I think we can make suspend side effects denote whether they perform blocking I/O or not and prevent compilation from happening when you mix the wrong effect types with the wrong CoroutinesContext.

raulraja commented on modules/docs/arrow-docs/docs/docs/effects/fx/README.md:
```diff
@@ -4,183 +4,553 @@ title: Fx
-fx {
-  sayHello() // Does not compile because Arrow Fx forces uses to denote side effects in `effect` blocks
+The `greet` program is ready to run as soon as the user is ready to commit to an execution strategy that is either `blocking` or `non-blocking`.
+`blocking` execution strategies will block the current thread waiting for the program to yield a value whereas `non-blocking` strategies will immediately return and perform the program's work without blocking the current thread.
```
We can guarantee it runs non-blocking by pushing the program execution to a Fiber started by the default dispatcher. I will block a thread and potentially a CPU in the worst case but it won't be the one where it started. True that is based on implementation but so is the entire effects runtime. We can't formally verify implementations are actually non-blocking

nomisRev replied:
Alright, that's perfect!

raulraja commented on modules/docs/arrow-docs/docs/docs/effects/fx/README.md:
```diff
@@ -4,183 +4,553 @@ title: Fx
+
+## Asynchronous & Concurrent Programming
+
+Arrow Fx benefits from auto-binding and direct syntax for asynchronous programming yielding extremely succinct programs without callbacks. This allow us to use direct style syntax with asynchronous and concurrent operations while preserving effect control in the types and runtime.
```
In a sense yes, I think we need to rename it to non-blocking binding or similar.

raulraja commented:
@pakoito @nomisRev addressed all your comments and separated doc into 3 docs. Intro, async + concurrency, and polymorphism. These were the 3 bigger sections. From this docs we are creating a post, @MaureenElsberry is helping me with that and also they will be enriched with drawings that @rafaparadela is helping with. The docs will be better in a couple of weeks.

nomisRev commented on modules/docs/arrow-docs/docs/docs/effects/fx/README.md:
```diff
@@ -4,183 +4,553 @@ title: Fx
+Arrow Fx brings purity, referential transparency and direct imperative syntax to typed FP in Kotlin.
 
-Arrow Fx opens a world of purity and controlled effects to help you describe your favorite effectful programs in both polymorphic and concrete styles as you do today using a wrapper like IO, Deferred or Observable but with direct syntax and unwrapped values that respect non-blocking semantics.
+Arrow Fx programs run unmodified in multiple supported frameworks and runtimes such as Arrow Effects IO, KotlinX Coroutines Deferred, Rx2 Observable and many more.
```
We should try but since it's not there yet won't it be confusing to have it in the docs?

JorgeCastilloPrz replied:
From a feature point of view, what's the benefit of providing a Deferred type? Can't you achieve the same stuff with IO already?

I thought the only existing purpose for DeferredK was to wrap the kotlinx.coroutines one.

nomisRev commented on modules/docs/arrow-docs/docs/docs/effects/fx/README.md:
```diff
@@ -4,183 +4,553 @@ title: Fx
+import arrow.effects.extensions.io.unsafeRun.runBlocking
+import arrow.effects.extensions.io.fx.fxCancellable
+//sampleStart
+val (_, disposable) = fxCancellable {
```
@pakoito @raulraja I just checked `bindingCancellable` again but I am a little confused.

Does this not break referential transparency if the `disposable` is a global handler.

```kotlin
fun main() {
  val (program, disposable) = IO.concurrent().fxCancelable({
    IO.unit.bind()
    IO { println("Hello I am here") }.bind()
  })

  program.fix().unsafeRunSync() //prints Hello I am here
  disposable.invoke()
  program.fix().unsafeRunSync() //Exception in thread "main" BindingCancellationException(message=null)

}
```

nomisRev replied:
Alright makes sense

raulraja commented:
@nomisRev added https://github.com/arrow-kt/arrow/issues/1281 and included it in the docs where it was referenced.
